go/merkletree2/tree.go

```diff
@@ -388,6 +437,25 @@ type MerkleInclusionProof struct {
 	RootMetadataNoHash  RootMetadata `codec:"e"`
 }
 
+// A MerkleExtensionProof proves, given the RootMetadata hashes of two merkle
+// trees and their respective Seqno values, that: - the two merkle trees have
+// the expected Seqno values, - the most recent merkle tree "points back" to the
+// least recent one through a chain of SkipPointers that refer to merkle trees
+// at intermediate Seqnos.
+type MerkleExtensionProof struct {
```

Comment:
In general, I'd like to discuss the strategy behind our protocol data structures. I think using toarray without versions is very brittle and to be avoided. I prefer instead maps with very short keys. See sig3/prot for an example of what I think is a better strategy. This allows us to extend in the future in ways we don't anticipate. The problem with arrays is that new fields inserted can break old clients in horrible ways. You can potentially protect everything with a version, but that's more work upfront.

Reply:
There's a little bit of extra overhead for maps with short keys, which might be one possible reason to use that strategy in very bandwidth-sensitive places.

Reply:
I think our chat conversation took care of this. Marking as resolved for now, feel free to reopen is not.